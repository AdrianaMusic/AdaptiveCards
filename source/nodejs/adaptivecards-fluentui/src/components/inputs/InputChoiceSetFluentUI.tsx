// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as AC from "adaptivecards";
import * as FluentUI from "@fluentui/react";
import * as Shared from "../../utils/shared";

export class InputChoiceSetFluentUI extends AC.ChoiceSetInput {

	private _value;
	public get value(): any {
	  return this._value || this.defaultValue;
	}
  
	public isSet(): any {
	  return this._value !== undefined || this.defaultValue !== undefined;
	}

	private _selectedValues: string[] = [];

	
	protected internalRender(): HTMLElement {
        const element = Shared.sharedInternalRender(this.renderReact);
        element.style.width = "100%";
        return element;
    }


    protected renderReact = (): JSX.Element => (
        this.isCompact ?
            <this.ComboBox /> :
            this.createExpandedView()
    )

    private parseChoises = (json: any) => {
        this.choices = Array.isArray(json.choices) ?
            json.choices.map(ch => {
                let choice = new AC.Choice();
                choice.parse(ch);
                return choice;
            }) :
            [];
    }

    private ComboBox = (): JSX.Element => (
        <FluentUI.ComboBox
            onChange={this.handleComboBoxChange}
            id={this.id}
            options={this.inputChoicesToComboBoxOptions(this.choices)}
            multiSelect={this.isMultiSelect}
			defaultSelectedKey={this.defaultValue}
			defaultValue={this.defaultValue}
			multiSelectDelimiter=","
            selectedKey={this._value}
        />
    )

    private createExpandedView = (): JSX.Element => {
        return this.isMultiSelect ?
            this.inputChoicesToCheckbox() :
            this.createChoiceGroup();
    }

    private createChoiceGroup = (): JSX.Element => (
        <FluentUI.ChoiceGroup
            id={this.id}
            onChange={this.handleChoiceGroupChange}
            options={this.inputChoicesToChoiceGroupOptions(
                this.choices,
                this.defaultValueToArray(this.defaultValue))}
            multiple={this.isMultiSelect}
        />
    )

    private inputChoicesToCheckbox = (): JSX.Element => (
        <React.Fragment>
            {this.choices.map((c, i) =>
                <FluentUI.Checkbox
                    key={c.value}
                    label={c.title}
                    defaultChecked={this.defaultValueToArray(this.value).includes(c.value)}
                    onChange={(ev, checked) => {
                        this.updateMultiselectData(checked, c.value);
                        ev.stopPropagation();
                        ev.preventDefault();
                    }}
                    styles={{
                        root: {
                            paddingTop: i > 0 ? this.hostConfig.spacing.default : 0,
                        },
                    }}
                />,
            )}
        </React.Fragment>
    )

    private defaultValueToArray = (value: string): string[] => {
        if (value) {
            return value.split(",");
        }
        return [];
    }

    private inputChoicesToComboBoxOptions = (choices: AC.Choice[]): FluentUI.IComboBoxOption[] => {
        return choices.map<FluentUI.IComboBoxOption>(c => ({ key: c.value, text: c.title }));
    }

    private inputChoicesToChoiceGroupOptions = (choices: AC.Choice[], value: string[]): FluentUI.IChoiceGroupOption[] => {
        return choices.map<FluentUI.IChoiceGroupOption>(c => ({ key: c.value, text: c.title, checked: value.includes(c.value) }));
    }

    private handleChoiceGroupChange =
        (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, option?: FluentUI.IChoiceGroupOption): void => {
            this._value = option.key;
            ev.stopPropagation();
            ev.preventDefault();
        }

    private handleComboBoxChange =
        (event: React.FormEvent<FluentUI.IComboBox>, option?: FluentUI.IComboBoxOption, index?: number, value?: string): void => {
            if (this.isMultiSelect) {
                this.updateMultiselectData(option.selected, option.key);
            } else {
                this._value = `${option.key}`;
            }
            event.stopPropagation();
            event.preventDefault();
        }

    private updateMultiselectData = (selected: boolean, key: any): void => {
        this.updateSelectedValues(selected, key);
        this._value = this._selectedValues.join(",");
    }

    private updateSelectedValues = (selected: boolean, key: any): void => {
        selected ? this._selectedValues.push(key) : this.removeItemFromArray(this._selectedValues, key);
    }

    private removeItemFromArray = (arr: any[], item: any): void => {
        const index = arr.indexOf(item);
        arr.splice(index, 1);
    }
}
