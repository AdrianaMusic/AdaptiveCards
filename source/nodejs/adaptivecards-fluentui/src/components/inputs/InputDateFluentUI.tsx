// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as FluentUI from "@fluentui/react";
import * as AC from "adaptivecards";
import * as Shared from "../../utils/shared";

export class InputDateFluentUI extends AC.DateInput {
  private handleSelectDate = (date: Date) => {
    this._value = date
      .toLocaleString(window.navigator.language, {
        year: "numeric",
        month: "2-digit",
        day: "2-digit",
      })
      .replace(/(\d+)\/(\d+)\/(\d+)/, "$3-$1-$2");
  };

  private _value;
  public get value(): any {
    return this._value || this.defaultValue;
  }

  public isSet(): any {
    return this._value !== undefined || this.defaultValue !== undefined;
  }

  protected internalRender(): HTMLElement {
    const element = Shared.sharedInternalRender(this.renderReact);

    element.style.width = "100%";
    return element;
  }

  private getDate = (
    dateString: string,
    errors?: AC.IValidationEvent[]
  ): Date => {
    try {
      return new Date(dateString + "T00:00:00");
    } catch (error) {
      Shared.raiseParseError(
        {
          phase: AC.ValidationPhase.Parse,
          event: AC.ValidationEvent.InvalidPropertyValue,
          message: error,
        },
        errors
      );
      return undefined;
    }
  };

  // NOT WORKING
  private onFormatDate = (date?: Date): string => {
    return !date
      ? ""
      : date.toLocaleString(window.navigator.language, { weekday: "short" }) +
          " " +
          (date.getMonth() + 1) +
          "/" +
          date.getDay() +
          "/" +
          date.getFullYear();
  };

  private onParseDateFromString = (val: string): Date => {
    const date = this.value || new Date();
    // const values = (val || '').trim().split('/');
    // const day = val.length > 0 ? Math.max(1, Math.min(31, parseInt(values[0], 10))) : date.getDate();
    // const month = val.length > 1 ? Math.max(1, Math.min(12, parseInt(values[1], 10))) - 1 : date.getMonth();
    // let year = val.length > 2 ? parseInt(values[2], 10) : date.getFullYear();
    // if (year < 100) {
    //   year += date.getFullYear() - (date.getFullYear() % 100);
    // }
    return new Date(this.defaultValue);
    //return new Date(year, month, day);
  };

  private renderReact = (): JSX.Element => {
    return (
      <FluentUI.DatePicker
        id={this.id}
        onSelectDate={this.handleSelectDate}
        value={this.defaultValue ? this.getDate(this.defaultValue) : undefined}
        //label={this.label}
        placeholder={this.placeholder}
        //formatDate={this.onFormatDate}
        //allowTextInput={true}
        parseDateFromString={this.onParseDateFromString}
        minDate={this.min ? this.getDate(this.min) : undefined}
        maxDate={this.max ? this.getDate(this.max) : undefined}
      />
    );
  };
}
