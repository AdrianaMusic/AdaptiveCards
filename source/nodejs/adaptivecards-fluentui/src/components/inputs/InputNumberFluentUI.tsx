// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as AC from "adaptivecards";
import * as FluentUI from "@fluentui/react";
import * as Shared from "../../utils/shared";

export class InputNumberFabric extends AC.NumberInput {

	protected handleChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
		this._value = newValue;
    }

	private _value;
	public get value(): any
	{
		return this._value || this.defaultValue;
	}

	public isSet(): any {
		return this._value !== undefined || this.defaultValue !== undefined;
	}

	protected internalRender(): HTMLElement {
        const element = Shared.sharedInternalRender(this.renderReact);
        element.style.width = "100%";
        return element;
    }

    protected renderReact = (): JSX.Element => (
        <FluentUI.TextField
            id={this.id}
            placeholder={`${this.placeholder}`}
            defaultValue={`${this.defaultValue}`}
            type="number"
            max={this.max}
            min={this.min}
            onChange={this.handleChange}
        />
    )

}
