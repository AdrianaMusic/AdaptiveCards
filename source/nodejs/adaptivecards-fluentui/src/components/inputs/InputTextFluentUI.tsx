// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as FluentUI from "@fluentui/react";
import * as AC from "adaptivecards";
import * as Shared from "../../utils/shared";

export class InputTextFluentUI extends AC.TextInput {
	protected handleChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
		this._value = newValue;
    }

	private _value;
	public get value(): any
	{
		return this._value || this.defaultValue;
	}

	public isSet(): any {
		return this._value !== undefined || this.defaultValue !== undefined;
	}
	
	protected internalRender(): HTMLElement {
        const element = Shared.sharedInternalRender(this.renderReact);
        element.style.width = "100%";
        return element;
    }

    private renderReact = (): JSX.Element => {
        return (
            <FluentUI.TextField
                id={this.id}
                type={AC.InputTextStyle[this.style].toLowerCase()}
				defaultValue={this.defaultValue}
				multiline={this.isMultiline}
                maxLength={this.maxLength ? this.maxLength : undefined}
                placeholder={this.placeholder}
				required={this.isRequired}
                //label={this.label}
                onChange={this.handleChange}
                onKeyDown={this.handleKeyDown}
            />
        );
    }

    private handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
        // Enter pressed
        if (e.keyCode === 13 && this.inlineAction) {
            this.inlineAction.execute();
        }
    }
}
