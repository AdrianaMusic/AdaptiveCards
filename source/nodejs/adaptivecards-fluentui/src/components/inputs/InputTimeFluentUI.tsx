// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as AC from "adaptivecards";
import * as FluentUI from "@fluentui/react";
import * as Shared from "../../utils/shared";

export class InputTimeFluentUI extends AC.TimeInput {

	protected internalRender(): HTMLElement {
        const element = Shared.sharedInternalRender(this.renderReact);
        element.style.width = "100%";
        return element;
    }

	protected handleChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
		this._value = newValue;
    }

	private _value;
	public get value(): any
	{
		return this._value || this.defaultValue;
	}

	public isSet(): any {
		return this._value !== undefined || this.defaultValue !== undefined;
	}


    protected renderReact = (): JSX.Element => {
        return (
            <FluentUI.TextField
                id={this.id}
                placeholder={`${this.placeholder}`}
                defaultValue={`${this.defaultValue}`}
                type="time"
				//label={this.label}
				required={this.isRequired}
                onChange={this.handleChange}
                max={this.max}
                min={this.min}
            />
        );
    }
}
