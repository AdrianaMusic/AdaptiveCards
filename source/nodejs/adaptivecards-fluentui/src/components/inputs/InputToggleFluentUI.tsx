// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as AC from "adaptivecards";
import * as FluentUI from "@fluentui/react";
import * as Shared from "../../utils/shared";

export class InputToggleFluentUI extends AC.ToggleInput {

	private handleToggleChange = (event: React.MouseEvent<HTMLElement> | React.ChangeEvent, checked?: boolean) => {
        this._value = checked ? this.valueOn : this.valueOff;
    }
	
	private _value;
	public get value(): any {
		return this._value;
	}

	public isSet(): any {
		return this._value ? true : false;
	}
	
	protected internalRender(): HTMLElement {
		if(this.defaultValue) {
			this._value = this.defaultValue;
		}
		
        const element = Shared.sharedInternalRender(this.renderReact);
        element.style.width = "100%";
        return element;
    }

    protected renderReact = (): JSX.Element => (
        <FluentUI.Toggle
            id={this.id}
            inlineLabel={true}
            onChange={this.handleToggleChange}
            defaultChecked={this.defaultValue === this.valueOn}
            label={this.title}
            styles={{
                root: {
                    marginBottom: 0,
                    marginTop: 0,
                },
            }}
        />
    )

 
}
