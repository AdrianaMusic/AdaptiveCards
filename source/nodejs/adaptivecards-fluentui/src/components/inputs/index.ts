export * from "./InputChoiceSetFluentUI";
export * from "./InputDateFluentUI";
export * from "./InputNumberFluentUI";
export * from "./InputTextFluentUI";
export * from "./InputTimeFluentUI";
export * from "./InputToggleFluentUI";
