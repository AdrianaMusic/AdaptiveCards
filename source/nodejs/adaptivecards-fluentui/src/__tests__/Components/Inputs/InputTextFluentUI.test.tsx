// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { InputTextFluentUI } from "../../../components/inputs/InputTextFluentUI";
import * as TestUtils from "../../TestUtils";
import { initializeIcons } from "@fluentui/react";

initializeIcons();

describe("InputTextExtended", () => {
    it("should be rendered successfully", () => {
        let input = new InputTextFluentUI();

        expect(input).toBeDefined();
        expect(input.render()).toBeDefined();
    });

    it("renders int text multiline correctly", () => {
        let input = TestUtils.renderFluentComponent(InputTextFluentUI, {
            "type": "Input.Text",
            "id": "input3",
            "placeholder": "enter comment",
            "maxLength": 500,
            "isMultiline": true,
            "value": "This value was pre-filled",
        });

        expect(input).toMatchSnapshot();
    });

    it("renders int text single line correctly", () => {
        let input = TestUtils.renderFluentComponent(InputTextFluentUI, {
            "type": "Input.Text",
            "id": "input3",
            "placeholder": "enter comment",
            "maxLength": 500,
            "isMultiline": false,
            "value": "This value was pre-filled",
        });

        expect(input).toMatchSnapshot();
    });

    it("can serialize inline action", () => {
        let input = new InputTextFluentUI();
        input.parse({
            "type": "Input.Text",
            "id": "input1",
            "inlineAction": {
                "type": "Action.Submit",
                "iconUrl": "http://adaptivecards.io/content/cats/1.png",
                "title": "Reply",
            },
        });

        let json = input.toJSON();

        expect(json).toBeDefined();
        expect(json.inlineAction).toBeDefined();
    });

    it("renders text box with inline action correctly", () => {
        let input = TestUtils.renderFluentComponent(InputTextFluentUI, {
            "type": "Input.Text",
            "id": "input3",
            "placeholder": "enter comment",
            "maxLength": 500,
            "isMultiline": false,
            "value": "This value was pre-filled",
        });

        expect(input).toMatchSnapshot();
    });

    it("can set initial value", () => {
        let el = TestUtils.renderFluentComponent(InputTextFluentUI, {
            "type": "Input.Text",
            "id": "input3",
            "placeholder": "enter comment",
            "maxLength": 500,
            "value": "This value was pre-filled",
        });

        expect(el).toBeDefined();
        let inputElement = el.querySelector("input");
        expect(inputElement).not.toBe(null);
        expect(inputElement.getAttribute("value")).toBe("This value was pre-filled");
    });

    it("can render placeholder", () => {
        let el = TestUtils.renderFluentComponent(InputTextFluentUI, {
            "type": "Input.Text",
            "id": "blah",
            "placeholder": "enter comment",
        });

        expect(el).toBeDefined();
        let inputElement = el.querySelector("input");
        expect(inputElement).not.toBe(null);
        expect(inputElement.getAttribute("placeholder")).toBe("enter comment");
    });

    it("can get json name", () => {
        expect(new InputTextFluentUI().getJsonTypeName()).toBe("Input.Text");
    });

    it("generates json correctly", () => {
        const element = new InputTextFluentUI();
        element.parse({
            "type": "Input.Text",
            "id": "blah",
            "placeholder": "enter comment",
            "style": "url"
        });

        const json = element.toJSON();
        expect(json).toBeDefined();
        expect(json.type).toEqual("Input.Text");
        expect(json.placeholder).toEqual("enter comment");
        expect(json.style).toEqual("Url");
    });
});
