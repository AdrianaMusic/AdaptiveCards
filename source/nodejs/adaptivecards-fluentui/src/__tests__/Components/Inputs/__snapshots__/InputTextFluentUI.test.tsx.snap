// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTextExtended renders int text multiline correctly 1`] = `
<div
  id="input3"
  style="display: flex; flex-direction: column; box-sizing: border-box; flex: 0 0 auto;"
>
  <div
    class="ac-input-container"
    style="display: flex;"
  >
    <div
      id="__ac-ecd3bb5a-2e0a-4b57-bf67-b4baa95410b2"
      style="width: 100%; min-width: 0px;"
    >
      <div
        class="ms-TextField ms-TextField--multiline root-41"
      >
        <div
          class="ms-TextField-wrapper"
        >
          <div
            class="ms-TextField-fieldGroup fieldGroup-52"
          >
            <textarea
              aria-invalid="false"
              class="ms-TextField-field field-53"
              id="input3"
              maxlength="500"
              placeholder="enter comment"
              type="text"
            >
              This value was pre-filled
            </textarea>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`InputTextExtended renders int text single line correctly 1`] = `
<div
  id="input3"
  style="display: flex; flex-direction: column; box-sizing: border-box; flex: 0 0 auto;"
>
  <div
    class="ac-input-container"
    style="display: flex;"
  >
    <div
      id="__ac-1d5b6e87-1ccc-40e3-9a6a-02a4bcdc8a96"
      style="width: 100%; min-width: 0px;"
    >
      <div
        class="ms-TextField root-41"
      >
        <div
          class="ms-TextField-wrapper"
        >
          <div
            class="ms-TextField-fieldGroup fieldGroup-42"
          >
            <input
              aria-invalid="false"
              class="ms-TextField-field field-43"
              id="input3"
              maxlength="500"
              placeholder="enter comment"
              type="text"
              value="This value was pre-filled"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`InputTextExtended renders text box with inline action correctly 1`] = `
<div
  id="input3"
  style="display: flex; flex-direction: column; box-sizing: border-box; flex: 0 0 auto;"
>
  <div
    class="ac-input-container"
    style="display: flex;"
  >
    <div
      id="__ac-4ebb0091-2bd2-4807-b523-ac916446b539"
      style="width: 100%; min-width: 0px;"
    >
      <div
        class="ms-TextField root-41"
      >
        <div
          class="ms-TextField-wrapper"
        >
          <div
            class="ms-TextField-fieldGroup fieldGroup-42"
          >
            <input
              aria-invalid="false"
              class="ms-TextField-field field-43"
              id="input3"
              maxlength="500"
              placeholder="enter comment"
              type="text"
              value="This value was pre-filled"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
